patches/coreboot-4.8.1/0000-measuredboot.patch

```diff
@@ -0,0 +1,474 @@
++	int i;
++
++	sha1_update(ctx, (uint8_t *)"\x80", 1);
++	while ((ctx->count % sizeof(ctx->buf)) != (sizeof(ctx->buf) - 8))
++		sha1_update(ctx, (uint8_t *)"\0", 1);
++
++	for (i = 0; i < 8; ++i) {
++		uint8_t tmp = cnt >> ((7 - i) * 8);
++		sha1_update(ctx, &tmp, 1);
++	}
++
++	for (i = 0; i < 5; i++)
++		ctx->buf.w[i] = __builtin_bswap32(ctx->state[i]);
++
++	return ctx->buf.b;
++}
++
++void sha1_init(struct sha1_ctx *ctx)
++{
++	ctx->state[0] = 0x67452301;
++	ctx->state[1] = 0xEFCDAB89;
++	ctx->state[2] = 0x98BADCFE;
++	ctx->state[3] = 0x10325476;
++	ctx->state[4] = 0xC3D2E1F0;
++	ctx->count = 0;
++}
+diff --git ./src/security/tpm/sha1.h ./src/security/tpm/sha1.h
+new file mode 100644
+index 0000000..e7e28e6
+--- /dev/null
++++ ./src/security/tpm/sha1.h
+@@ -0,0 +1,31 @@
++/* Copyright (c) 2014 The Chromium OS Authors. All rights reserved.
++ * Use of this source code is governed by a BSD-style license that can be
++ * found in the LICENSE file.
++ */
++
++/* SHA-1 functions */
++
++#ifndef _sha1_h_
++#define _sha1_h_
++
++#include <stdint.h>
++#include <commonlib/helpers.h>
++
++#define SHA1_DIGEST_SIZE 20
++#define SHA1_BLOCK_SIZE 64
++
++/* SHA-1 context */
++struct sha1_ctx {
++	uint32_t count;
++	uint32_t state[5];
++	union {
++		uint8_t b[SHA1_BLOCK_SIZE];
++		uint32_t w[DIV_ROUND_UP(SHA1_BLOCK_SIZE, sizeof(uint32_t))];
++	} buf;
++};
++
++void sha1_init(struct sha1_ctx *ctx);
++void sha1_update(struct sha1_ctx *ctx, const uint8_t *data, uint32_t len);
++uint8_t *sha1_final(struct sha1_ctx *ctx);
++
++#endif  /* _sha1_h_ */
+diff --git ./src/security/tpm/tss.h ./src/security/tpm/tss.h
+index 8f3f1cb..5c569cb 100644
+--- ./src/security/tpm/tss.h
++++ ./src/security/tpm/tss.h
+@@ -147,6 +147,11 @@ uint32_t tlcl_extend(int pcr_num, const uint8_t *in_digest,
+ 		     uint8_t *out_digest);
+ 
+ /**
++ * Perform a SHA1 hash on a region and extend a PCR with the hash.
++ */
++uint32_t tlcl_measure(int pcr_num, const void * start, size_t len);
++
++/**
+  * Get the entire set of permanent flags.
+  */
+ uint32_t tlcl_get_permanent_flags(TPM_PERMANENT_FLAGS *pflags);
+diff --git ./src/security/tpm/tss/tcg-1.2/tss.c ./src/security/tpm/tss/tcg-1.2/tss.c
+index 161d29f..4577ec4 100644
+--- ./src/security/tpm/tss/tcg-1.2/tss.c
++++ ./src/security/tpm/tss/tcg-1.2/tss.c
+@@ -17,6 +17,7 @@
+ #include <arch/early_variables.h>
+ #include <assert.h>
+ #include <string.h>
++#include <security/tpm/sha1.h>
+ #include <security/tpm/tis.h>
+ #include <vb2_api.h>
+ #include <security/tpm/tss.h>
+@@ -354,3 +355,23 @@ uint32_t tlcl_extend(int pcr_num, const uint8_t *in_digest,
+ 		       kPcrDigestLength);
+ 	return result;
+ }
++
++
++uint32_t tlcl_measure(int pcr_num, const void * start, size_t len)
++{
++	VBDEBUG("TPM: pcr %d measure %p @ %zu: ", pcr_num, start, len);
++
++	struct sha1_ctx sha;
++	sha1_init(&sha);
++	sha1_update(&sha, start, len);
++
++	const uint8_t * hash = sha1_final(&sha);
++	for(unsigned i = 0 ; i < SHA1_DIGEST_SIZE ; i++)
```

Comment:
I would remove the dumping of the hash from here, it is just making it easy for someone to see how the PCRs were extending by simply doing a 'cbmem'. So let's not put that in any 'production code' and comment out this loop.

Reply:
@osresearch what do you think? I don't see any harm in enabling debug since the contents of what goes into the PCR is not sensitive.